### PR TITLE
📦️ First batch of weights

### PR DESCRIPTION
**Pull request by lrazovic**
## What and Why?

It's crucial to know the computational resources needed for different functions, including runtime functions like `on_initialize` and `verify_unsigned`, to maintain runtime safety and manage transaction inclusion based on available resources. We can determine this by running benchmarks on reference hardware, which will provide the weights for each runtime function.

> Weight is two-dimensional, it tracks both time (execution time) and space (state accesses). Weight determines how much fees need to be paid in order to perform some operation.

## How?
Using an AWS EC2 `c7a.4xlarge` instance, we can run the benchmarks in an isolated, standardized, and replicable environment. The CPU, particularly in single-core performance, meets the minimum requirements outlined in the Knowledge Hub and on the [Polkadot Wiki](https://wiki.polkadot.network/docs/maintain-guides-how-to-validate-polkadot#reference-hardware).

To benchmark a specific pallet/runtime, you can use the command: `just benchmark-runtime polimec-local <pallet_name>`.

## Anything Else?

Since we're currently using weights generated for a different runtime, it would be wise to double-check the outputs concerning the weights actually in use, which can be found [here](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame). We should also compare these with the weights used in system chains like those on [Asset Hub](https://github.com/polkadot-fellows/runtimes/tree/main/system-parachains/asset-hubs/asset-hub-polkadot/src/weights), to spot if we have major discrepancies.

### CODE REVIEW

File: runtimes/polimec/src/lib.rs [hunk 1]
```diff
@@ -723,7 +723,7 @@ impl pallet_parachain_staking::Config for Runtime {
 	type RewardPaymentDelay = RewardPaymentDelay;
 	type RuntimeEvent = RuntimeEvent;
 	type RuntimeHoldReason = RuntimeHoldReason;
-	type WeightInfo = pallet_parachain_staking::weights::SubstrateWeight<Runtime>;
+	type WeightInfo = ();
```
Comment (lrazovic) [inline runtimes/polimec/src/lib.rs:726]: We need to adjust the benchmarks. This will be addressed in an upcoming PR.

**Review by vstam1 — APPROVED** [review]
Lgtm. Small question (is a my mistake). Currently all the settlement extrinsics do not have the correct weight set (just a Weight of 0). Could you use the correct weight benchmarking there in this PR?

File: runtimes/polimec/src/lib.rs [hunk 2]
```diff
@@ -893,7 +893,7 @@ impl pallet_identity::Config for Runtime {
 	type Slashed = Treasury;
 	type SubAccountDeposit = SubAccountDeposit;
 	type UsernameAuthorityOrigin = EnsureNever<AccountId>;
-	type WeightInfo = pallet_identity::weights::SubstrateWeight<Runtime>;
+	type WeightInfo = ();
```
[the remark this answers is not in the document]
Comment (lrazovic) [reply runtimes/polimec/src/lib.rs:896]: Fixed here https://github.com/Polimec/polimec-node/pull/231

File: runtimes/polimec/src/lib.rs (hunk 1 above)
Comment (lrazovic) [reply runtimes/polimec/src/lib.rs]: Fixed in https://github.com/Polimec/polimec-node/pull/235.

File: runtimes/polimec/src/lib.rs [hunk 3]
```diff
@@ -534,7 +534,7 @@ impl pallet_treasury::Config for Runtime {
 	type SpendFunds = ();
 	type SpendOrigin = frame_support::traits::NeverEnsureOrigin<Balance>;
 	type SpendPeriod = SpendPeriod;
-	type WeightInfo = pallet_treasury::weights::SubstrateWeight<Runtime>;
+	type WeightInfo = ();
```
[the remark this answers is not in the document]
Comment (lrazovic) [reply runtimes/polimec/src/lib.rs]: Fixed in https://github.com/Polimec/polimec-node/pull/233.

**Review by JuaniRios — CHANGES_REQUESTED** [review]
Can you add screenshots with the weight diffs of each changed weight file?
You can fix your previous issue by renaming both files you want to compare to the same name.
Would save time for me and Just if one person does that once

**Comment by lrazovic** [comment]
> Can you add screenshots with the weight diffs of each changed weight file? You can fix your previous issue by renaming both files you want to compare to the same name. Would save time for me and Just if one person does that once

Sure. [Here](https://gist.github.com/lrazovic/170b2d112b7343a0adda2ad31ba033e2) you can find a list showcasing the main differences between the weights we were using before, and the new ones. The comparison is not exactly optimal because `subweight` sometimes has problems comparing weights generated with different runtimes, as in this case, but the next time we have to generate weights, the comparison will be more reliable.

**Comment by lrazovic** [comment]
We realized that the `WeightInfo` for `pallet_xcm` are missing, the benchmarks needs a fix :( @vstam1 

```
Error: Input("Benchmark pallet_xcm::send failed: SendFailure")
2024-04-15 18:03:00 Starting benchmark: pallet_xcm::send
```

Command to reproduce: `cargo run --features runtime-benchmarks --release benchmark pallet --chain=polimec-local --pallet="pallet_xcm" --extrinsic="*"`

**Comment by lrazovic** [comment]
> We realized that the `WeightInfo` for `pallet_xcm` are missing, the benchmarks needs a fix :( @vstam1
> 
> ```
> Error: Input("Benchmark pallet_xcm::send failed: SendFailure")
> 2024-04-15 18:03:00 Starting benchmark: pallet_xcm::send
> ```
> 
> Command to reproduce: `cargo run --features runtime-benchmarks --release benchmark pallet --chain=polimec-local --pallet="pallet_xcm" --extrinsic="*"`

Fixed in https://github.com/Polimec/polimec-node/pull/239.

**Comment by JuaniRios** [comment]
CleanShot 2024-04-17 at 11 41 46
`just dry-run-benchmarks` is failing on pallet-funding

**Review by JuaniRios — APPROVED** [review]
Seems the heaviest changes are between 20 and 80%. It's hard to tell if it's too much since we are comparing pallet weights with a real runtime weights.

Let's merge for now, and keep in mind the next weight diffs should be <20% (Joe's feedback)

**Pull request merged**